state/endpoint_bindings.go

```diff
@@ -398,10 +411,132 @@ func DefaultEndpointBindingsForCharm(charmMeta *charm.Meta) map[string]string {
 	allRelations := charmMeta.CombinedRelations()
 	bindings := make(map[string]string, len(allRelations)+len(charmMeta.ExtraBindings))
 	for name := range allRelations {
-		bindings[name] = network.DefaultSpaceName
+		bindings[name] = network.DefaultSpaceId
 	}
 	for name := range charmMeta.ExtraBindings {
-		bindings[name] = network.DefaultSpaceName
+		bindings[name] = network.DefaultSpaceId
 	}
 	return bindings
 }
+
+//go:generate mockgen -package mocks -destination mocks/endpointbinding_mock.go github.com/juju/juju/state EndpointBinding
+type EndpointBinding interface {
+	SpaceByID(id string) (*Space, error)
+	SpaceIDsByName() (map[string]string, error)
+	SpaceNamesByID() (map[string]string, error)
+}
+
+type Bindings struct {
+	st  EndpointBinding
+	app *Application
+	bindingsMap
+}
+
+// NewBindings returns a bindings guaranteed to be in space id format.
+func NewBindings(st EndpointBinding, givenMap map[string]string) (*Bindings, error) {
```

Comment:
This needs work, not happy with the way it came out.  Wanted to get the rest of it up for review overnight.

Reply:
improved.